Sources/PrimerSDK/Classes/Core/Primer/DependencyInjection.swift

```diff
@@ -36,36 +40,47 @@ final internal class DependencyContainer {
 
     private func register<T>(_ dependency: T) {
         let key = String(describing: T.self)
-        dependencies[key] = dependency as AnyObject
+        Self.queue.async(flags: .barrier) {
+            self.dependencies[key] = dependency as AnyObject
+        }
     }
 
     private func resolve<T>() -> T {
         let key = String(describing: T.self)
-        let dependency = dependencies[key] as? T
-
-        if dependency == nil {
-            if key == String(describing: AppStateProtocol.self) {
-                let appState: AppStateProtocol = AppState()
-                DependencyContainer.register(appState)
-                return self.resolve()
-
-            } else if key == String(describing: PrimerSettingsProtocol.self) {
-                let primerSettings: PrimerSettingsProtocol = PrimerSettings()
-                DependencyContainer.register(primerSettings)
-                return self.resolve()
-
-            } else if key == String(describing: PrimerThemeProtocol.self) {
-                let primerTheme: PrimerThemeProtocol = PrimerTheme()
-                DependencyContainer.register(primerTheme)
-                return self.resolve()
-            }
+
+        if let dependency = Self.queue.sync(execute: { dependencies[key] as? T }) {
+            return dependency
         }
 
-        precondition(
-            dependency != nil,
-            "No dependency found for \(key)! must register a dependency before resolve."
-        )
+        return Self.queue.sync {
```

Comment:
We could use only one queue.sync here.

Reply:
```
return Self.queue.sync {
            if let dependency = self.dependencies[key] as? T {
                return dependency
            }

            let dependency: T? = self.createDependency(for: key)
            if let dependency = dependency {
                self.dependencies[key] = dependency as AnyObject
                return dependency
            }

            preconditionFailure("No dependency found for \(key)! Must register a dependency before resolve.")
        }
```
something like this?

Reply:
Did you try to run Unit Tests with this change? It was crashing for me (deadlock, I think) with one queue.
Trying once more to make sure it wasnt a red herring

Reply:
I did not try tests... leave it as it is, it is a nitpick anyway.